### PR TITLE
Lots of Hybrid Improvements

### PR DESCRIPTION
**Pull request by varunagrawal**
1. `HybridGaussianFactorGraph::discreteFactors()` to retrieve all the discrete factors. More of a convenience method.
2. `prune` method for `HybridNonlinearFactor`.
3. Prune in `reorderRelinearize` to improve nonlinear efficiency.
4. Improve Switching tests.

### CODE REVIEW

File: gtsam/hybrid/HybridGaussianFactorGraph.cpp [hunk 1]
```diff
@@ -580,4 +580,15 @@ GaussianFactorGraph HybridGaussianFactorGraph::choose(
   return gfg;
 }
 
+/* ************************************************************************ */
+DiscreteFactorGraph HybridGaussianFactorGraph::discreteFactors() const {
+  DiscreteFactorGraph dfg;
+  for (auto &&f : factors_) {
+    auto discreteFactor = std::dynamic_pointer_cast<DiscreteFactor>(f);
+    assert(discreteFactor);
```
Comment (dellaert) [inline gtsam/hybrid/HybridGaussianFactorGraph.cpp]: I don’t understand how this can work. Not all factors are discrete, right?

**Review by dellaert — COMMENTED** [review]
Just one question

File: gtsam/hybrid/HybridGaussianFactorGraph.cpp (hunk 1 above)
Comment (varunagrawal) [reply gtsam/hybrid/HybridGaussianFactorGraph.cpp]: Yup, which is why we have the assert, but I guess the `if(auto discreteFactor = ...)` check is better.

**Review by dellaert — APPROVED** [review]
OK, that makes more sense :-)

**Pull request merged**